geth/mailservice/mailservice_api.go

```diff
@@ -16,52 +18,130 @@ import (
-	From uint32 `json:"from"`
-
-	// To is a upper bound of time range (optional).
-	// Default is now.
-	To uint32 `json:"to"`
+func (api *PublicAPI) provider() ServiceProvider {
+	return api.service.provider
+}
 
-	// Topic is a regular Whisper topic.
-	Topic whisper.TopicType `json:"topic"`
+// addPeer tries to connect with a peer and waits till connection will be established.
+func (api *PublicAPI) addPeer(peer *discover.Node, timeout time.Duration) error {
+	log.Debug("adding a peer", "peer", peer.String())
+	server, err := api.provider().Server()
+	if err != nil {
+		return err
+	}
+	events := make(chan *p2p.PeerEvent, 10)
+	sub := server.SubscribeEvents(events)
+	defer sub.Unsubscribe()
+	server.AddPeer(peer)
+	for {
+		select {
+		case ev := <-events:
+			if ev.Type == p2p.PeerEventTypeAdd && ev.Peer == peer.ID {
+				log.Debug("peer added", "peer", peer.String())
+				return nil
+			}
+		case <-time.After(timeout):
+			return fmt.Errorf("failed to add a peer: %s", peer.String())
+		}
+	}
+}
 
-	// SymKeyID is an ID of a symmetric key to authenticate to MailServer.
-	// It's derived from MailServer password.
-	SymKeyID string `json:"symKeyID"`
+// runTrustedPeersGC collects connections that weren't used for defined
+// garbage collector timeout.
+func (api *PublicAPI) runTrustedPeersGC(timeout, period time.Duration) {
+	connectedPeers := map[*discover.Node]time.Time{}
+	ticker := time.NewTicker(period)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-api.service.quit:
+			log.Debug("shutdown mailserver api gc")
+			return
+		case peer := <-api.newConnectedPeers:
+			connectedPeers[peer] = time.Now()
+		case <-ticker.C:
+			for peer, lastUsed := range connectedPeers {
+				if time.Since(lastUsed) >= timeout {
+					server, err := api.provider().Server()
+					// node was stopped
+					if err != nil {
+						log.Debug("exit from mailserver api gc. failed to get a server", "error", err)
+						return
+					}
+					server.RemovePeer(peer)
+					delete(connectedPeers, peer)
+				}
+			}
+		}
+	}
 }
 
-func setMessagesRequestDefaults(r *MessagesRequest) {
-	// set From and To defaults
-	if r.From == 0 && r.To == 0 {
-		r.From = uint32(time.Now().UTC().Add(-24 * time.Hour).Unix())
-		r.To = uint32(time.Now().UTC().Unix())
+// choosePeer loops over trusted nodes, and returns one that is connected or
+// tries to establish a connection.
+func (api *PublicAPI) choosePeer() (*discover.Node, error) {
+	server, err := api.provider().Server()
+	if err != nil {
+		return nil, err
+	}
+	if len(server.Config.TrustedNodes) == 0 {
+		return nil, ErrNoServersInConfig
+	}
+	// we are not relying on GC for this to avoid any mismatch between
+	// real data and GC. GC used only to disconnect peers that didn't
+	// disconnect themself
+	connected := map[discover.NodeID]struct{}{}
+	for _, peer := range server.Peers() {
+		connected[peer.ID()] = struct{}{}
+	}
+	for _, trusted := range server.Config.TrustedNodes {
+		if _, exist := connected[trusted.ID]; exist {
```

Comment:
It looks like there a chance that we can select a peer that is being removed at the same time.